spacy_curated_transformers/models/listeners.py

```diff
@@ -254,82 +336,93 @@ def __init__(
-            dX = model._backprop(dX, outputs_to_backprop=outputs_to_backprop)
+            ListenerStateUtils.verify_inputs(model, docs)
 
-            model._batch_id = None
-            model._outputs = None
-            model._backprop = None
+            Y, backprop_pooling = pooling(_outputs.all_outputs, is_train)
 
-            return dX
+            def backprop(dY):
+                dX = backprop_pooling(dY)
 
-        return Y, backprop
-    else:
-        width = model.get_dim("nO")
+                if grad_factor != 1.0:
+                    for dX_doc in dX:
+                        for dX_layer in dX_doc:
+                            dX_layer.data *= grad_factor
 
-        no_trf_data = [doc._.trf_data is None for doc in docs]
-        if any(no_trf_data):
-            assert all(no_trf_data)
-            return [
-                [model.ops.alloc2f(len(doc), width) for _ in range(n_layers)]
-                for doc in docs
-            ], lambda dY: []
+                outputs_to_backprop = tuple(i for i in range(0, _outputs.num_outputs))
+                dX = _backprop(dX, outputs_to_backprop=outputs_to_backprop)
 
-        if any(doc._.trf_data.last_layer_only for doc in docs):
-            raise ValueError(
-                Errors.E012.format(listener_name="TransformerLayersListener")
-            )
+                ListenerStateUtils.clear_state(model)
+                return dX
 
-        return pooling.predict(docs), lambda dY: []
+            return Y, backprop
 
+        else:
+            if ListenerStateUtils.use_doc_annotations_for_prediction(model):
+                width = model.get_dim("nO")
+
+                no_trf_data = [doc._.trf_data is None for doc in docs]
+                if any(no_trf_data):
+                    assert all(no_trf_data)
+                    return [
+                        [model.ops.alloc2f(len(doc), width) for _ in range(n_layers)]
+                        for doc in docs
+                    ], lambda dY: []
+
+                if any(doc._.trf_data.last_layer_only for doc in docs):
+                    raise ValueError(
+                        Errors.E012.format(listener_name="TransformerLayersListener")
+                    )
+
+                return pooling.predict(docs), lambda dY: []
+            else:
+                assert _outputs is not None
+                ListenerStateUtils.verify_inputs(model, docs)
+                outputs: Tuple[List[List[Floats2d]], Callable[[Any], Any]] = (
+                    pooling.predict(_outputs.all_outputs),
+                    lambda dY: [],
+                )
+                ListenerStateUtils.clear_state(model)
+                return outputs
 
-class LastTransformerLayerListener(TransformerListener):
-    """Extracts the output of the last transformer layer and performs pooling over the
-    individual pieces of each Doc token, returning their corresponding representations.
-    """
 
+class LastTransformerLayerListener:
```

Comment:
I was looking at #7 again, but I think that (besides the `ListenerStateUtils`) all these classes with static methods are quite intuitive. If I am not missing anything, we could eg. inline `LastTransformerLayerListener.new` in the (registry) function that constructs the model and then make standalone function for forward, eg. `LastTransformerLayerListener.forward` -> `last_layer_listener_forward`.

Reply:
Done.